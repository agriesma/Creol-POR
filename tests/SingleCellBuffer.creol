/* SingleCellBuffer.creol

   by Einar Broch Johnson */

interface Buff
begin
op put(in x : Data)
op get(out x : Data)
end

class Buffer implements Buff
begin
var cell : Data
var ctr : Int
var next : Buff

op init == ctr := 0;
op run == skip;
op put(in x : Data) ==
    if ctr = 0 then
      cell := x ;
    else
      if next = null then
	next := new Buffer();
      fi
      next.put(x;);
    fi
    ctr := ctr + 1;

op get(out x : Data) ==
    await (ctr > 0) ; 
    if cell /= null then
      x := cell;
    else
      next.get(; x);
    fi
    ctr := ctr - 1;
    cell := null;

end
