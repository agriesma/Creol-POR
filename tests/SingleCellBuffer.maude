load interpreter
mod PROGRAM is
pr CREOL-INTERPRETER .
op init : -> Configuration [ctor] .
eq init =
< "Buffer" : Cl | Inh: noInh, Par: noAid, Att: "cell" |-> null , "ctr" |-> null , "next" |-> null, Mtds: 
  < "init" : Mtdname | Param: noAid, Latt: noSubst, Code: "ctr" ::= int(0) ; return ( emp ) > *
  < "run" : Mtdname | Param: noAid, Latt: noSubst, Code: skip ; return ( emp ) > *
  < "put" : Mtdname | Param: "x", Latt: "x" |-> null, Code: if "=" ( "ctr" # int(0) ) th "cell" ::= "x" el if "=" ( "next" # null ) th "next" ::= new "Buffer" ( emp ) el skip fi ; "sync.0" ! "next" . "put" ( "x" ) ; ( "sync.0" ? ( noAid ) )  fi ; "ctr" ::= "+" ( "ctr" # int(1) ) ; return ( emp ) > *
  < "get" : Mtdname | Param: noAid, Latt: "x" |-> null, Code: ( await ">" ( "ctr" # int(0) ) ) ; if "/=" ( "cell" # null ) th "x" ::= "cell" el "sync.0" ! "next" . "get" ( emp ) ; ( "sync.0" ? ( "x" ) )  fi ; "ctr" ::= "-" ( "ctr" # int(1) ) ; "cell" ::= null ; return ( "x" ) >, Ocnt: 0 >

.
endm
