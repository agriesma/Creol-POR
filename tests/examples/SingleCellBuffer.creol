/*
 * SingleCellBuffer.creol -- Dynamic buffer example.
 *
 * This file is part of creoltools
 *
 * Written and Copyright (c) 2007 by Einar Broch Johnsen
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

interface Buff
begin
with Any
  op put(in x : Data)
  op get(out x : Data)
end

class Buffer implements Buff
begin
  var cell : Data
  var ctr : Int
  var next : Buff

  op init == ctr := 0
  op run == skip
  with Any op put(in x : Data) ==
    if ctr = 0 then cell := x
    else if next = null then next := new Buffer() end; next.put(x;) end;
    ctr := ctr + 1

  with Any op get(out x : Data) ==
    await (ctr > 0) ;
    if cell /= null then
      x := cell
    else
      next.get(; x)
    end;
    ctr := ctr - 1;
    cell := null

end
