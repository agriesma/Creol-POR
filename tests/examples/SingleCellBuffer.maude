load creol-interpreter
mod PROGRAM is
protecting CREOL-SIMULATOR .
op init : -> Configuration [ctor] .
eq init =
< "Buffer" : Cl | Inh: noInh, Par: noVid, Att: "cell" |-> null , "ctr" |-> null , "next" |-> null, Mtds: 
  < "init" : Mtdname | Param: noVid, Latt: noSubst, Code: "ctr" ::= int(0) ; return ( emp ) > *
  < "run" : Mtdname | Param: noVid, Latt: noSubst, Code: skip ; return ( emp ) > *
  < "put" : Mtdname | Param: "x", Latt: "x" |-> null , "label:0" |-> null, Code: if "=" ( "ctr" :: int(0) ) th "cell" ::= "x" el if "=" ( "next" :: null ) th "next" ::= new "Buffer" ( emp ) el skip fi ; "label:0" ! "next" . "put" ( "x" ) ; ( "label:0" ? ( noVid ) )  fi ; "ctr" ::= "+" ( "ctr" :: int(1) ) ; return ( emp ) > *
  < "get" : Mtdname | Param: noVid, Latt: "x" |-> null , "label:0" |-> null, Code: ( await ">" ( "ctr" :: int(0) ) ) ; if "/=" ( "cell" :: null ) th "x" ::= "cell" el "label:0" ! "next" . "get" ( emp ) ; ( "label:0" ? ( "x" ) )  fi ; "ctr" ::= "-" ( "ctr" :: int(1) ) ; "cell" ::= null ; return ( "x" ) >, Ocnt: 0 >

.
endm
