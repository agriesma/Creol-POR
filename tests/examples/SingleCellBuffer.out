==========================================
reduce in PROGRAM : init .
result Class: 
< "Buffer" : Cl | Inh: noInh, Par: noVid, Att: "cell" |-> null, "ctr" |-> null,
    "next" |-> null, Mtds:
  < "get" : Mtdname | Param: noVid, Latt: "label:0" |-> null, "x" |-> null,
    Code: await ">"("ctr" # int(0)) ; if "/="("cell" # null) th "x" ::= "cell"
    el "label:0" ! "next" . "get"(emp) ; "label:0" ?("x") fi ; "ctr" ::= "-"(
    "ctr" # int(1)) ; "cell" ::= null ; return("x") > *
  < "init" : Mtdname | Param: noVid, Latt: noSubst, Code: "ctr" ::= int(0) ;
    return(emp) > *
  < "put" : Mtdname | Param: "x", Latt: "label:0" |-> null, "x" |-> null, Code:
    if "="("ctr" # int(0)) th "cell" ::= "x" el if "="("next" # null) th "next"
    ::= new "Buffer"(emp) el skip fi ; "label:0" ! "next" . "put"("x") ;
    "label:0" ?(noVid) fi ; "ctr" ::= "+"("ctr" # int(1)) ; return(emp) > *
  < "run" : Mtdname | Param: noVid, Latt: noSubst, Code: skip ; return(emp) >,
  Ocnt: 0 >
Bye.
